### PR TITLE
Fix/hardcoded date tenancy service

### PR DESCRIPTION
**Pull request by danielwilliams94**
This PR addresses a potential reliability flaw in the current setup with regards to retrieving new tenancies.

With the existing setup, the controller will always retrieve tenancies that have been added in the past day. It is relying on the OutSystems scheduler running once a day to make sure that none are missed. This is potentially problematic as if the scheduler fails or is even delayed by a matter of minutes, there is a significant possibility that some tenancies will be missed.

The fix for this is that we are persisting the timestamp of the last time this method was run in a database table. When it is run again, it retrieves all values that have been added since that timestamp. If no timestamp exists, it falls back on the previous 24 hour period.

We have added a new DBContext within a new DB project (this can be extended in the future). We have added a repository and a service.

### CODE REVIEW

File: ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs [hunk 1]
```diff
@@ -10,34 +10,159 @@
...
+            _mockClock = new Mock<IClock>();
             _mockHousingApiCall = new Mock<IHackneyHousingAPICall>();
+            _mockLastRetrieved = new Mock<INewTenancyService>();
             var mockHousingApiCallBuilder = new Mock<IHackneyHousingAPICallBuilder>();
             var mockCrmToken = new Mock<IHackneyGetCRM365Token>();
-            _service = new TenancyService(mockHousingApiCallBuilder.Object, _mockHousingApiCall.Object, mockCrmToken.Object);
+            _service = new TenancyService(mockHousingApiCallBuilder.Object, _mockHousingApiCall.Object, mockCrmToken.Object, _mockLastRetrieved.Object, _mockClock.Object);
+        }
+
+        private static HttpResponseMessage CreateResponseMessage(Dictionary<string, object> response)
```
Comment (PinkhamSoftware) [inline ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs]: Not sure this needs to be static :) Resharper will often refactor methods and make it static.

File: ManageATenancyAPI.Tests/Unit/Services/NewTenancyServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            
+            Assert.Equal(yesterday, result);
+        }
+
+        [Fact]
+        public void UpdateLastRetrieved_CallsRepository()
+        {
+            _tenancyRepository.Setup(m => m.UpdateLastRun(It.IsAny<DateTime>())).Verifiable();
+            
+            _service.UpdateLastRetrieved(new DateTime());
```
Comment (PinkhamSoftware) [inline ManageATenancyAPI.Tests/Unit/Services/NewTenancyServiceTests.cs]: Would have liked to see an exact date go in and get verified, if possible.

File: ManageATenancyAPI/Helpers/Clock.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+using System;
+using ManageATenancyAPI.Interfaces;
+
+namespace ManageATenancyAPI.Helpers
+{
+    public class Clock : IClock
+    {
+        public DateTime Now => DateTime.Now;
```
Comment (danielwilliams94) [inline ManageATenancyAPI/Helpers/Clock.cs]: Use UTCNow here

File: ManageATenancyAPI/Repository/TenancyRepository.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        }
+        
+        public NewTenancyLastRunDto GetLastRun()
+        {
+            var lastRun = _tenancyContext.NewTenancyLastRun.FirstOrDefault();
+            return lastRun == null ? null : new NewTenancyLastRunDto(lastRun);
+        }
+
+        public void UpdateLastRun(DateTime dateTime)
+        {
```
Comment (danielwilliams94) [inline ManageATenancyAPI/Repository/TenancyRepository.cs]: Change this to actually store a running history of runtimes

File: ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs [hunk 5]
```diff
@@ -10,34 +10,159 @@
...
+        public async Task GetNewTenancies_UsesLastRunDate()
+        {
+            var lastRunTime = new DateTime(1995, 03, 15, 22, 55, 33, 21);
+            var queryDate = lastRunTime.ToCrmQueryFormat();
+
+            var calledQuery = "";
+            _mockHousingApiCall.Setup(x =>
+                    x.getHousingAPIResponse(It.IsAny<HttpClient>(), It.IsAny<string>(), It.IsAny<string>()))
+                .Callback((HttpClient client, string query, string parameter) => { calledQuery = query; })
+                .ReturnsAsync(CreateResponseMessage(new Dictionary<string, object>()));
```
Comment (danielwilliams94) [inline ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs]: Refactor this into a method with a clear name

Comment (danielwilliams94) [reply ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs]: Give the calledquery a better name like "Dynamic365Query"

File: ManageATenancyAPI/Helpers/Clock.cs (hunk 3 above)
Comment (danielwilliams94) [reply ManageATenancyAPI/Helpers/Clock.cs]: Done

File: ManageATenancyAPI/Repository/TenancyRepository.cs (hunk 4 above)
Comment (danielwilliams94) [reply ManageATenancyAPI/Repository/TenancyRepository.cs]: Fixed this

File: ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs (hunk 5 above)
Comment (danielwilliams94) [reply ManageATenancyAPI.Tests/Unit/Services/TenancyServiceTests.cs]: Done.

**Pull request merged**